ext/pdo/pdo_stmt.c

```diff
@@ -1266,12 +1266,10 @@ PHP_METHOD(PDOStatement, fetchObject)
 
 	do_fetch_opt_finish(stmt, 0);
 
-	if (ctor_args) {
```

Comment:
Originally, I proposed an `UNKNOWN` default value because I thought that it's intentional that `$ctorArgs` has different behavior when:
- it's not passed at all (nothing happens)
- it's empty (`VAL_UNDEF(&stmt->fetch.cls.ctor_args);`)
- it's not empty (`ZVAL_ARR(&stmt->fetch.cls.ctor_args, zend_array_dup(Z_ARRVAL_P(ctor_args)));`)

To be honest, it's still not clear for me unfortunately what the intented behavior is, I couldn't provoke any issues (which hasn't existed yet) when I made the ZPP nullable. So any insights are welcome about what the preferred fix is.

Reply:
Yeah, I can see that the implementation is a little messy. The idea behind this method is that you can mix and match it with other fetch methods and the default fetch mode should always be retained. For example, code like this should be valid:
```php
class A
{
    public function __construct(...$arr)
    {
        var_dump($arr);
    }
}

$stmt->setFetchMode(PDO::FETCH_CLASS, A::class, ['123']);

$a1 = $stmt->fetchObject(A::class, []);
$std = $stmt->fetchObject(null, []);
$a2 = $stmt->fetch();
```
The default is created with ctor arguments, but you don't have to use `fetch` or iterator everytime. You can use `fetchObject` in between. The way that PDO is designed, the values would get overwritten, which is why there is this weird logic in `fetchObject`. Apparently, it does not work well because I am getting segmentation faults.

Reply:
I pushed a fix for the memory leak, but I'm still not sure what we should be doing with this code. It seems pointless to support null for arguments if it's going to do the same as an empty array.

Reply:
It looks like we do support null for setFetchMode, but again null is treated the same way as empty array.

Reply:
Thank you!

> It seems pointless to support null for arguments if it's going to do the same as an empty array

Yes, I had the same thought yesterday.. If we are strict about BC then we should accept this fact, in my opinion :( Although other constructor arguments are already non-nullable (I believe I made some related changes in the past).

P.S. I've just noticed that we slightly fcuked up some "constructor args" parameter names. E.g.: mysqli and pgsql use `$constructor_args`, while PDO uses `$ctorArgs` :D

Reply:
Well, we have a funny case here. I think for `fetchObject()`, it would actually be fine to change the signature to `array $ctorArgs = []`. The reason is that `?array $ctorArgs = null` is considered compatible with that (it accepts more inputs), while at the same time, the implementation didn't actually accept null, so can't be breaking code that *is* passing null either. So I think this change is actually backwards-compatible, in an odd way ^^

My only concern with not allow null is that setFetchMode() allows it -- but as you mention, we also have pg_fetch_object() and mysqli_fetch_object() which both don't allow null for constructor args, so not allowing it here would be consistent with that.

Reply:
> P.S. I've just noticed that we slightly fcuked up some "constructor args" parameter names. E.g.: mysqli and pgsql use $constructor_args, while PDO uses $ctorArgs :D

Hm ... as the new name is not documented yet (it's still `$ctor_args` in the docs right now), maybe we should change it (would be `$constructorArgs`)? Though I don't particularly care about this, it's a minor difference, not the case of outright wrong name.

Reply:
> The reason is that ?array $ctorArgs = null is considered compatible with that (it accepts more inputs), while at the same time, the implementation didn't actually accept null, so can't be breaking code that is passing null either. So I think this change is actually backwards-compatible, in an odd way ^^

I would have never realized this, but you are absolutely right! LSP can be a bless sometimes :D That said, I'm happy to remove nullability.

> Hm ... as the new name is not documented yet (it's still $ctor_args in the docs right now), maybe we should change it (would be $constructorArgs)? Though I don't particularly care about this, it's a minor difference, not the case of outright wrong name.

Yeah, it's just a slight consistency issue, not like the other ones which really needed fixes. However, I'll create a PR to change the param name to `$constructorArgs`.